File: 0198-house-robber/0198-house-robber.cpp
class Solution {
public:
    int rob(vector<int>& nums) 
    {
        int n=nums.size();
        
        int curr{},pre{nums[0]},pre2{};
       
        int take{},nottake{};
        for(int i{1};i<n;i++)
        {
            take=nums[i];
            if(i>1)
            {
                take+=pre2;
            }
            nottake=pre;
            curr=max(take,nottake);
            pre2=pre;
            pre=curr;
            
        }
        return pre;
    }
};

File: 0198-house-robber/0198-house-robber.java
class Solution {
     int dp[]=new int [101];
    public int solve(int [] nums,int ind)
    {
        if(ind==0)
        {
            return nums[0];
        }
        if(ind<0)
        {
            return 0;
        }
        if(dp[ind]!=-1)
        {
            return dp[ind];
        }
        int pick=nums[ind]+solve(nums,ind-2);
        int notpick=solve(nums,ind-1);
        return dp[ind]=Math.max(pick,notpick);
    }
    public int rob(int[] nums)
    {
        int n=nums.length;
       
        Arrays.fill(dp,-1);
        return solve(nums,n-1);
        
    }
}